internal/transports/p2p/peer/checkpoint.go

```diff
@@ -0,0 +1,139 @@
+package peer
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/bitcoin-sv/block-headers-service/domains"
+	"github.com/bitcoin-sv/block-headers-service/internal/chaincfg"
+	"github.com/bitcoin-sv/block-headers-service/internal/chaincfg/chainhash"
+	"github.com/rs/zerolog"
+)
+
+type checkpoint struct {
+	checkpoint      *chaincfg.Checkpoint
+	currentIndex    int
+	finalCheckpoint *chaincfg.Checkpoint
+	checkpoints     []chaincfg.Checkpoint
+	log             *zerolog.Logger
+	lock            sync.RWMutex
+}
+
+func newCheckpoint(checkpoints []chaincfg.Checkpoint, tipHeight int32, log *zerolog.Logger) *checkpoint {
+	logger := log.With().Str("subservice", "checkpoint").Logger()
+	ch := &checkpoint{
+		checkpoints: checkpoints,
+		log:         &logger,
+	}
+
+	if len(checkpoints) != 0 {
+		ch.finalCheckpoint = &checkpoints[len(checkpoints)-1]
+		ch.next(tipHeight)
+	}
+
+	return ch
+}
+
+func (ch *checkpoint) Height() int32 {
+	ch.lock.RLock()
+	defer ch.lock.RUnlock()
+
+	return ch.checkpoint.Height
+}
+
+func (ch *checkpoint) Hash() *chainhash.Hash {
+	ch.lock.RLock()
+	defer ch.lock.RUnlock()
+
+	return ch.checkpoint.Hash
+}
+
+// LastReached returns true if the last checkpoint has been reached.
+func (ch *checkpoint) LastReached() bool {
+	ch.lock.RLock()
+	defer ch.lock.RUnlock()
+
+	return ch.checkpoint == nil
+}
+
+// Check checks if the header is valid according to the checkpoint and marks switches to next checkpoint if reached.
+func (ch *checkpoint) Check(header *domains.BlockHeader) error {
+	ch.lock.Lock()
+	defer ch.lock.Unlock()
+
+	if ch.checkpoint == nil {
+		return nil
+	}
+
+	if header.Height < ch.checkpoint.Height {
+		return nil
+	}
+
+	if header.Height == ch.checkpoint.Height {
+		if header.Hash != *ch.checkpoint.Hash {
+			return fmt.Errorf("corresponding checkpoint height does not match, got: %v, exp: %v", header.Height, ch.checkpoint.Height)
+		}
+
+		ch.next(header.Height)
+		return nil
+	}
+
+	if header.Height > ch.checkpoint.Height {
+		return fmt.Errorf("unexpected header above next checkpoint height, got: %v, for checkpoint at height %d", header, ch.checkpoint.Height)
+	}
+
+	return nil
+}
+
+// UseTip sets the next checkpoint to the valid checkpoint for given tip height.
+func (ch *checkpoint) UseTip(height int32) {
+	ch.lock.Lock()
+	defer ch.lock.Unlock()
+
+	ch.checkpoint = nil
+	ch.currentIndex = -1
+	ch.next(height)
+}
```

Comment:
I don't really see the point of using this method. We don't need to reset `checkpoint` and `currentIndex`, because they are immediately set in the `next` method. 
And I think it makes more sense to just add locking to `next` method and use it in `StartSyncHeaders` instead of this method.

Reply:
This is preparation for the case when we have multiple live connections but only one is syncing.
So imagine the case, when we're instantiating 8 peers, 
one is syncing and we're for example in the middle of the checkpoint list.
Then connection is dropped and we're switching to other peer. But he's checkpoint is outdated, so we need to actually set it at the point of starting sync. 
Maybe better solution would be to create this component only at the moment of StartingSync not at the moment of creation of the peer, like it is currently, WDYT?

Reply:
But creating a new checkpoint at the moment of StartinSync is also I good idea IMO

Reply:
method removed, and checkpoint is created at the start sync method